### DISCUSSION TITLE
Plausible/Analytics does not compile with Elixir 1.15

### DISCUSSION THREAD

**Discussion by jnowls** [plausible/analytics#3156] [Self-Hosted Support] [answered]
### Past Issues Searched

- [X] I have searched open and closed issues to make sure that the bug has not yet been reported

### Issue is a Bug Report

- [X] This is a bug report and not a feature request, nor asking for self-hosted support

### Using official Plausible Cloud hosting or self-hosting?

Self-hosting

### Describe the bug

`mix deps.compile` fails due to `idna` not being a direct dependency of the project:

From the [Elixir 1.15 release notes](https://elixirforum.com/t/elixir-v1-15-0-released/56584):

> Due to the code path pruning, if you have an application or dependency
that does not specify its dependencies on Erlang and Elixir application,
it may no longer compile successfully in Elixir v1.15. You can temporarily
disable code path pruning by setting prune_code_paths: false in your
mix.exs, although doing so may lead to runtime bugs that are only
manifested inside a mix release.

As such, when trying to compile:
```
== Compilation error in file lib/public_suffix.ex ==
** (UndefinedFunctionError) function :idna.encode/2 is undefined (module :idna is not available)
    :idna.encode(~c"ac", [uts46: true])
    lib/public_suffix/rules_parser.ex:66: PublicSuffix.RulesParser.punycode_domain/1
    lib/public_suffix/rules_parser.ex:35: anonymous fn/1 in PublicSuffix.RulesParser.parse_rules_section/2
    (elixir 1.15.0) lib/stream.ex:507: anonymous fn/3 in Stream.flat_map/2
    (elixir 1.15.0) lib/stream.ex:990: Stream.do_transform_user/6
    (elixir 1.15.0) lib/enum.ex:4387: Enum.split_with/2
    lib/public_suffix/rules_parser.ex:38: PublicSuffix.RulesParser.parse_rules_section/2
    lib/public_suffix/rules_parser.ex:18: PublicSuffix.RulesParser.parse_rules/1
could not compile dependency :public_suffix, "mix compile" failed. Errors may have been logged above. You can recompile this dependency with "mix deps.compile public_suffix --force", update it with "mix deps.update public_suffix" or clean it with "mix deps.clean public_suffix"
```

I'm not an Elixir developer, I am just trying to get this application up and running outside of docker, however following the instructions in the release notes, adding `prune_code_paths: false` to my `mix.exs` seems to let `mix deps.compile` run successfully.

### Expected behavior

From my inexperienced perspective, the application should probably let `mix deps.compile` work without having to specifically disable `prune_code_paths`.

### Screenshots

--

### Environment

```markdown
- OS: Linux
- Browser: N/A
- Browser Version: N/A
```

**Comment by ruslandoga** [comment, marked as answer]
👋 @jnowls 

I guess the current workaround would be to use the same Elixir and Erlang versions as in https://asdf-vm.com config https://github.com/plausible/analytics/blob/34fbc3d5bcc31a94f7ac98b4ae410c676ecbac1d/.tool-versions#L1-L2 or https://github.com/plausible/analytics/blob/34fbc3d5bcc31a94f7ac98b4ae410c676ecbac1d/Dockerfile#L5

**Reply by ruslandoga** [reply]
Anything else is practically untested.

**Reply by jnowls** [reply]
Thanks, there's another workaround as well I mentioned in my original post. I'm not really asking for support, just wanted to report that this is probably something to look into, as from my understanding, there's dependencies that aren't properly reporting its own dependencies causing issues.

**Reply by ruslandoga** [reply]
Indeed, but luckily at least some dependencies declare it correctly and it's included in the final release in Elixir v1.14.

```console
$ docker compose exec plausible cat releases/0.0.1/plausible.rel | grep idna
          {idna,"6.1.1",permanent},
```

Thank you for catching this issue early :) But it'll probably be handled only once the project moves to v1.15.

**Reply by ruslandoga** [reply]
For later, here's the reason for the error: https://github.com/axelson/publicsuffix-elixir/blob/89372422ab8b433de508519ef474e39699fd11ca/mix.exs#L18-L20

The author of that package deliberately removed `idna` from the applications list and that's where the compilation fails.
